File: 프로그래머스/3/92345. 사라지는 발판/사라지는 발판.cpp
#include <string>
#include <vector>
#include <climits>

using namespace std;

struct Result {
    bool win;
    int moves;
};

int n, m;
int dx[4] = {1, 0, -1, 0};
int dy[4] = {0, 1, 0, -1};

// 현재 움직일 사람: (cx, cy), 상대: (ox, oy)
Result dfs(vector<vector<int>>& board, int cx, int cy, int ox, int oy) {
    if (board[cx][cy] == 0) return {false, 0};
    
    bool hasMove = false;

    int bestWin = INT_MAX; // 이길 수 있는 경우 중 최소 이동 수
    int bestLose = 0;      // 질 수밖에 없는 경우 중 최대 이동 수
    
    for (int dir = 0; dir < 4; dir++) {
        int nx = cx + dx[dir];
        int ny = cy + dy[dir];
        if (nx < 0 || nx >= n || ny < 0 || ny >= m) continue;
        if (board[nx][ny] == 0) continue;
        
        hasMove = true;
        
        board[cx][cy] = 0;
        Result child = dfs(board, ox, oy, nx, ny);
        board[cx][cy] = 1; 
        
        if (child.win) {
            bestLose = max(bestLose, child.moves + 1);
        } else {
            bestWin = min(bestWin, child.moves + 1);
        }
    }
    
    if (!hasMove) return {false, 0};
    if (bestWin != INT_MAX) return {true, bestWin};
    return {false, bestLose};
}

int solution(vector<vector<int>> board, vector<int> aloc, vector<int> bloc) {
    n = board.size();
    m = board[0].size();
    
    Result ans = dfs(board, aloc[0], aloc[1], bloc[0], bloc[1]);
    return ans.moves;
}

File: 프로그래머스/3/92345. 사라지는 발판/사라지는 발판.java
class Solution {
    int h, w;
    int[][] board;
    
    int[] dr = {-1, 1, 0, 0};
    int[] dc = {0, 0, -1, 1};
    
    public int solution(int[][] board, int[] aloc, int[] bloc) {
        this.board = board;
        this.h = board.length;
        this.w = board[0].length;
        
        Result result = dfs(aloc[0], aloc[1], bloc[0], bloc[1]);
        
        return result.count;
    }
    
    private Result dfs(int ar, int ac, int br, int bc) {
        if (board[ar][ac] == 0) {
            return new Result(false, 0);
        }
        
        boolean canMove = false;
        boolean canWin = false;
        
        int minWinCount = Integer.MAX_VALUE;
        int maxLoseCount = 0;
        
        for (int d = 0; d < 4; d++) {
            int nr = ar + dr[d];
            int nc = ac + dc[d];
            
            if (!isValid(nr, nc)) continue;
            
            canMove = true;
            
            board[ar][ac] = 0;
            
            Result next = dfs(br, bc, nr, nc);
            
            board[ar][ac] = 1;
            
            if (!next.win) {
                canWin = true;
                minWinCount = Math.min(minWinCount, next.count + 1);
            } else if (!canWin) {
                maxLoseCount = Math.max(maxLoseCount, next.count + 1);
            }
        }
        
        if (!canMove) {
            return new Result(false, 0);
        }
        
        if (canWin) {
            return new Result(true, minWinCount);
        }
        
        return new Result(false, maxLoseCount);
    }
    
    private boolean isValid(int r, int c) {
        return r >= 0 && r < h && c >= 0 && c < w && board[r][c] == 1;
    }
    
    static class Result {
        boolean win;
        int count;
        
        Result(boolean win, int count) {
            this.win = win;
            this.count = count;
        }
    }
}